media-proxy/src/mesh/conn_rdma_rx.cc

```diff
@@ -0,0 +1,154 @@
+                    log::error("Failed to pass empty buffer to RDMA to receive into")
+                              ("buffer_address", buf)("error", fi_strerror(-err))
+                              (" ", kind2str(_kind));
+                    res = add_to_queue(buf);
+                    if (res != Result::success) {
+                        log::error("Failed to add buffer to queue")("error", result2str(res))
+                                  (" ", kind2str(_kind));
+                        break;
+                    }
+                }
+            } else {
+                break; // Exit the loop to avoid spinning on errors
+            }
+        }
+        // Wait for the buffer to become available
+        wait_buf_available();
+    }
+}
+
+/**
+ * @brief Handles the RDMA completion queue (CQ) events in a dedicated thread.
+ * 
+ * This function continuously monitors the CQ for completion events, processes
+ * completed buffers, and returns them to the buffer queue for reuse. It ensures
+ * efficient event handling by reading events in batches and sleeping briefly
+ * when no events are available to avoid busy waiting.
+ * 
+ * Key Steps:
+ * 1. Reads a batch of CQ entries.
+ * 2. For each entry:
+ *    - Processes the received buffer and sends it.
+ *    - Adds the buffer back to the queue upon successful processing.
+ *    - Logs errors for any issues during processing or queue operations.
+ * 3. Handles `EAGAIN` errors by yielding CPU time to avoid busy looping.
+ * 4. Logs and exits on other CQ read errors.
+ * 
+ * @param ctx The context for managing thread cancellation and operations.
+ */
+void RdmaRx::rdma_cq_thread(context::Context& ctx)
+{
+    constexpr int CQ_RETRY_DELAY_US = 100; // Retry delay for EAGAIN
+    struct fi_cq_entry cq_entries[CQ_BATCH_SIZE]; // Array to hold batch of CQ entries
+
+    while (!ctx.cancelled()) {
+        // Read a batch of completion events
+        int ret = fi_cq_read(ep_ctx->cq_ctx.cq, cq_entries, CQ_BATCH_SIZE);
+        if (ret > 0) {
+            for (int i = 0; i < ret; ++i) {
+                void *buf = cq_entries[i].op_context;
+                if (buf == nullptr) {
+                    log::error("Null buffer context, skipping...")
+                              ("batch_index",i)(" ", kind2str(_kind));
+                    continue;
+                }
+
+                // Process the buffer (e.g., transmit or handle)
+                Result res = transmit(ctx, buf, trx_sz);
+                if (res != Result::success) {
+                    log::error("Failed to transmit buffer")("buffer_address", buf)("size", trx_sz)
+                              (" ", kind2str(_kind));
+                    continue;
+                }
+
+                // Add the buffer back to the queue
+                res = add_to_queue(buf);
+                if (res == Result::success) {
+                    // Notify that a buffer is available
+                    notify_buf_available();
```

Comment:
notify_buf_available() is already called inside add_to_queue()

Reply:
@moleksy Please include this in future refactoring.